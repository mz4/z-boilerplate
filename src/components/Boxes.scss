.Box {
  background-color: indigo;
  border: 1px solid #ccc;
  color: #f3f3f3;
  display: block;
  height: 300px;
  margin: 16px;
  padding: 16px;
  position:relative;
  width: 300px;
}

.Text {
  text-align: center;
  vertical-align: middle;
  line-height: 300px;
}

.Dropdown {
  position: relative;
  display: inline-block;
}

.dropbtn {
  background-color: #3498DB;
  color: white;
  padding: 16px;
  font-size: 16px;
  border: none;
  cursor: pointer;
}

.dropbtn:hover, .dropbtn:focus {
  background-color: #2980B9;
}
