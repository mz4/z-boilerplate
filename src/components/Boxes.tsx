import React, { useEffect, useRef, useState } from "react";
import Menu from "./Menu";
import "./Boxes.scss";

function useOnClickOutside(ref: any, handler: any) {
  useEffect(() => {
    const listener = (event: any) => {
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }
      handler(event);
    };
    document.addEventListener("mousedown", listener);
    document.addEventListener("touchstart", listener);
    return () => {
      document.removeEventListener("mousedown", listener);
      document.removeEventListener("touchstart", listener);
    };
  }, [ref, handler]);
}

const Boxes = () => {
  const [isMenuOpen, setMenuOpen] = useState(false);
  const menuRef = useRef(null);

  const openMenu = () => {
    setMenuOpen(!isMenuOpen);
  };

  useOnClickOutside(menuRef, () => {
    setMenuOpen(false);
  });

  return (
    <>
      <div className="Box" data-testid="boxes">
        <div className="Dropdown" ref={menuRef}>
          <button
            onClick={openMenu}
            className="dropbtn"
            data-testid="openMenuButton"
          >
            Menu
          </button>
          <Menu isMenuOpen={isMenuOpen} />
        </div>
        <div className="Text">Ciao</div>
      </div>
      <div className="Box">
        <div className="Text">Ciao</div>
      </div>
    </>
  );
};

export default Boxes;
