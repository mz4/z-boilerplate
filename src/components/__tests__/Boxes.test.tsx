import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Boxes from "../Boxes";

describe('Boxes', () => {
  it('should render component', () => {
    const {container} = render(<Boxes />);
    expect(container).toMatchSnapshot();
  });
  it('should render component', () => {
    const { queryByText, getByTestId } = render(<Boxes />);
    expect(queryByText('Home')).not.toBeInTheDocument();
    expect(queryByText('About')).not.toBeInTheDocument();
    expect(queryByText('Contact')).not.toBeInTheDocument();
    fireEvent.click(getByTestId("openMenuButton"));
    expect(queryByText('Home')).toBeInTheDocument();
    expect(queryByText('About')).toBeInTheDocument();
    expect(queryByText('Contact')).toBeInTheDocument();
  });
  it('should find multiple elements', () => {
    const { queryAllByText } = render(<Boxes />);
    const ciao = queryAllByText('Ciao');
    expect(ciao.length).toBe(2);
  });
});
