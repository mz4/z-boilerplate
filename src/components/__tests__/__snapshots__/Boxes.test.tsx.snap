// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boxes should render component 1`] = `
<div>
  <div
    class="Box"
    data-testid="boxes"
  >
    <div
      class="Dropdown"
    >
      <button
        class="dropbtn"
        data-testid="openMenuButton"
      >
        Menu
      </button>
      <div
        class="dropdown-content"
      />
    </div>
    <div
      class="Text"
    >
      Ciao
    </div>
  </div>
  <div
    class="Box"
  >
    <div
      class="Text"
    >
      Ciao
    </div>
  </div>
</div>
`;
